tools/ab_test.py

```diff
@@ -0,0 +1,260 @@
+#!/usr/bin/env python3
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""
+Script for running A/B-Tests
+
+The script takes two git revisions and a pytest integration test. It utilizes
+our integration test frameworks --binary-dir parameter to execute the given
+test using binaries compiled from each revision, and captures the EMF logs
+output. It the searches for list-valued properties in the EMF, and runs a
+regression test comparing these lists for the two runs.
+
+It performs the A/B-test as follows:
+For each EMF log message output, look at the dimensions. The script assumes that
+dimensions are unique across all log messages output from a single test run. In
+each log message, then look for all properties that have lists assigned to them,
+and collect them. For both runs of the test, the set of distinct dimensions
+collected this way must be the same. Then, we match corresponding dimensions
+between the two runs, performing statistical regression test across all the list-
+valued properties collected.
+"""
+import argparse
+import asyncio
+import json
+import platform
+import shutil
+import statistics
+import sys
+from pathlib import Path
+
+from aws_embedded_metrics.logger.metrics_logger_factory import create_metrics_logger
+
+# Hack to be able to use our test framework code
+sys.path.append(str(Path(__file__).parent.parent / "tests"))
+
+# pylint:disable=wrong-import-position
+from framework import utils
+from framework.ab_test import chdir, check_regression, git_ab_test
+
+
+def extract_dimensions(emf):
+    """Extracts the cloudwatch dimensions from an EMF log message"""
+    dimension_list = emf["_aws"]["CloudWatchMetrics"][0]["Dimensions"][0]
+    dimensions = {}
+
+    for key in emf:
+        if key in dimension_list:
+            dimensions[key] = emf[key]
```

Comment:
dict comprehensions a bit shorter and marginally clearer

```python
dimensions = {key: val for key, val in emf.items() if key in dimension_list}
```

Reply:
Y'know, I was kinda hoping you'd tell me there's a different solution to this altogether xP I'll take the dict comprs though :D

Reply:
Ah I get what you mean now... since we are just deconstrructing and EMF message and re-constructing it again... We could just send it directly, but from looking at the library it does not look possible.

One thing we could do is put the message directly to the agent:

```python
import socket
from urllib.parse import urlparse
emf_endpoint = urlparse(os.environ["AWS_EMF_AGENT_ENDPOINT"])
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.sendto(json.dumps(emf_msg), (emf_endpoint.hostname, emf_endpoint.port))
sock.close()
```